### PR TITLE
simplifying matrix combinations

### PR DESCRIPTION
**Pull request by pritidesai**
# Changes

Matrix combinations is created in `[][]Param` format which is very hard to process with matrix.Include. The parameters value can change in a combination or a new combination can be added based on `matrix.Include.Params`.

Its easy to compare and work with map[string]string instead of []Param.

Now, this combination can be directly consumed while processing `matrix.Include` as demonstrated here:  https://play.golang.com/p/Hy48eFFtlVo 

/kind cleanup

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/pipeline/v1/matrix_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1/matrix_types.go:73]: wondering if we can define `[]map[string]string` as `combinations` so that we can create member functions on it, for example:

 ```suggestion
 type combinations []map[string]string
 
func (c *combinations) distribute(param Param) combinations {
```

Comment (pritidesai) [reply pkg/apis/pipeline/v1/matrix_types.go:73]: thanks @jerop I can look into it. I have been talking to @EmmaMunley to utilize the same `Matrix` struct instead of creating a new `struct` to hold all possible combinations for the given `matrix`. This new field in the existing struct will not be exposed to the users. Thoughts?

Comment (jerop) [reply pkg/apis/pipeline/v1/matrix_types.go:73]: sounds good 👍🏾

Comment (pritidesai) [reply pkg/apis/pipeline/v1/matrix_types.go:73]: I love to hear the significance of member functions 🎉 makes the code so much cleaner and maintainable 🎉

**Comment by pritidesai** [comment]
@EmmaMunley this is still WIP but just opened it up for an early review and to unblock your efforts, thanks!

File: pkg/apis/pipeline/v1beta1/matrix_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/matrix_types.go:72]:
```suggestion
func (c Combinations) Format() []Params {
```

to remove repetition

File: pkg/apis/pipeline/v1beta1/matrix_types.go
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/matrix_types.go:45]: why is combinations added to the struct?

File: pkg/apis/pipeline/v1beta1/matrix_types.go
Comment (pritidesai) [reply pkg/apis/pipeline/v1beta1/matrix_types.go]: sure, changed it to `ToParams`, doesn't avoid repetition though 🤣

File: pkg/apis/pipeline/v1beta1/matrix_types.go
Comment (pritidesai) [reply pkg/apis/pipeline/v1beta1/matrix_types.go:46]: such that all combinations are stored along with a given `matrix`. This facilitates calling `FanOut` once and store all the combinations for further reuse instead of calling `FanOut` again or passing `combinations` around.

Any cons of introducing this here?

File: docs/pipeline-api.md
Comment (pritidesai) [inline docs/pipeline-api.md:8661]: I am surprised to see this kind of github tags here? 🤔

Comment (pritidesai) [reply docs/pipeline-api.md:8661]: its coming from https://github.com/tektoncd/pipeline/blob/e8f26371d22a3420458d23c2684b12a961df9198/docs/pipeline-api.md#param-1

Comment (lbernick) [reply docs/pipeline-api.md:8661]: https://github.com/tektoncd/pipeline/issues/6210
I think we should prioritize this issue-- I will hopefully have some time to look at it

File: pkg/apis/pipeline/v1/matrix_types.go
Comment (lbernick) [inline pkg/apis/pipeline/v1/matrix_types.go:64]:
```suggestion
// Combinations holds a list of combination for a given Matrix
```

File: pkg/apis/pipeline/v1beta1/matrix_types.go
Comment (lbernick) [reply pkg/apis/pipeline/v1beta1/matrix_types.go:46]: I have the same comment as jerop-- I think it's a bit confusing for calling functions to have to know whether or not `FanOut` has been called yet to populate matrix combinations.

File: pkg/apis/pipeline/v1/matrix_types.go
Comment (pritidesai) [reply pkg/apis/pipeline/v1/matrix_types.go:119]: Storing the sorted combinations guarantees the same order and its deterministic - sorted based on param names.

We could sort in tests (`matrix_types_test.go` and `pipelinerun_test.go`) instead (also means sorting will be necessary in any tests we add in future) but I could not figure out how to sort `params` in `&taskRuns.Items[i]` without making it complicated, please feel free to suggest any fix.

https://github.com/tektoncd/pipeline/blob/55d217b6d8aab8ebb13fc54ad50f1e0cd18f97fe/pkg/reconciler/pipelinerun/pipelinerun_test.go#L8531

**Comment by pritidesai** [comment]
Hey @jerop and @lbernick, @EmmaMunley is waiting for this PR, will appreciate if we can expedite the review 🙏

**Comment by lbernick** [comment]
> Hey @jerop and @lbernick, @EmmaMunley is waiting for this PR, will appreciate if we can expedite the review 🙏

np! feel free to use the "request review" button (I do tend to check emails from GH) or DM me when you need another round

File: pkg/apis/pipeline/v1beta1/matrix_types.go
Comment (lbernick) [reply pkg/apis/pipeline/v1beta1/matrix_types.go:46]: I do think it would be less confusing to have Combinations separate from the matrix struct. I see in the reconciler code
```
rpt.PipelineTask.Matrix.FanOut()
matrixCombinations := rpt.PipelineTask.Matrix.Combinations.ToParams()
```

and I think 
```
combinations := rpt.PipelineTask.Matrix.FanOut().ToParams()
```
is clearer. (`FanOut` is not something I'd expect to be mutating, and `rpt.PipelineTask.Matrix.Combinations.ToParams()` would produce the wrong result if FanOut hadn't been called.)

File: pkg/apis/pipeline/v1/matrix_types.go
Comment (lbernick) [reply pkg/apis/pipeline/v1/matrix_types.go:119]: ok, sorting the params here sounds fine. I was a bit worried about people coming to depend on this functionality even if we don't guarantee it. Maybe we could just note in the docs that param ordering is not guaranteed?

Comment (pritidesai) [reply pkg/apis/pipeline/v1/matrix_types.go:114]: the ordering is only for the internal data structure `combinations` (`[]Params`) which is not exposed to the users.

Comment (lbernick) [reply pkg/apis/pipeline/v1/matrix_types.go:114]: ah ok, nvm!

File: pkg/apis/pipeline/v1/matrix_types.go
Comment (lbernick) [inline pkg/apis/pipeline/v1/matrix_types.go:59]:
```suggestion
// Combinations is a slice of Combinations
```

Suggesting this because combinations doesn't reference the matrix that created them

**Review by lbernick — APPROVED** [review]

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:876]: it seems you meant to delete this line, the function is actually used in 877 below
```suggestion

```

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (jerop) [inline pkg/reconciler/pipelinerun/pipelinerun.go:801]: it seems you meant to delete this line, the function is actually used in 802 below
```suggestion

```

**Review by jerop — COMMENTED** [review]
one small change requested, happy to merge after it and lee's comment is addressed ,thanks @pritidesai 🙏🏾

**Comment by pritidesai** [comment]
> one small change requested, happy to merge after it and lee's comment is addressed ,thanks @pritidesai 🙏🏾

thanks @jerop, both yours and @lbernick's comments addressed, thanks!

**Review by jerop — APPROVED** [review]
/lgtm

🎉

**Pull request merged**
